tfx/examples/bert_cola/bert_utils.py

```diff
@@ -0,0 +1,155 @@
+      tf.reshape(feature, [-1]),
+      max_len=_MAX_LEN)
+
+def preprocessing_fn(inputs):
+  """tf.transform's callback function for preprocessing inputs.
+
+  Args:
+    inputs: map from feature keys to raw not-yet-transformed features.
+
+  Returns:
+    Map from string feature key to transformed feature operations.
+  """
+  feature = inputs['feature']
+  label = inputs['label']
+  input_word_ids, input_mask, segment_ids = _tokenize(feature)
+  return {
+          'label': label,
+          'input_word_ids': input_word_ids,
+          'input_mask': input_mask,
+          'segment_ids': segment_ids
+          }
+
+def _input_fn(file_pattern: List[Text],
+              tf_transform_output: tft.TFTransformOutput,
+              batch_size: int = 200) -> tf.data.Dataset:
+  """Generates features and label for tuning/training.
+
+  Args:
+    file_pattern: List of paths or patterns of input tfrecord files.
+    tf_transform_output: A TFTransformOutput.
+    batch_size: representing the number of consecutive elements of returned
+      dataset to combine in a single batch
+
+  Returns:
+    A dataset that contains (features, indices) tuple where features is a
+      dictionary of Tensors, and indices is a single Tensor of label indices.
+  """
+  transformed_feature_spec = (
+      tf_transform_output.transformed_feature_spec().copy())
+
+  dataset = tf.data.experimental.make_batched_features_dataset(
+      file_pattern=file_pattern,
+      batch_size=batch_size,
+      features=transformed_feature_spec,
+      reader=_gzip_reader_fn,
+      label_key=_LABEL_KEY)
+
+  return dataset
+  
+def _get_serve_tf_examples_fn(model, tf_transform_output):
+  """Returns a function that parses a serialized tf.Example."""
+
+  model.tft_layer = tf_transform_output.transform_features_layer()
+
+  @tf.function
+  def serve_tf_examples_fn(serialized_tf_examples):
+    """Returns the output to be used in the serving signature."""
+    feature_spec = tf_transform_output.raw_feature_spec()
+    feature_spec.pop(_LABEL_KEY)
+    parsed_features = tf.io.parse_example(serialized_tf_examples, feature_spec)
+
+    transformed_features = model.tft_layer(parsed_features)
+
+    return model(transformed_features)
+
+  return serve_tf_examples_fn
+
+# TFX Trainer will call this function.
+def run_fn(fn_args: TrainerFnArgs):
+  """Train the model based on given args.
+
+  Args:
+    fn_args: Holds args used to train the model as name/value pairs.
+  """
+  tf_transform_output = tft.TFTransformOutput(fn_args.transform_output)
+
+  train_dataset = _input_fn(fn_args.train_files, tf_transform_output,
+                            batch_size=_TRAIN_BATCH_SIZE)
+  eval_dataset = _input_fn(fn_args.eval_files, tf_transform_output,
+                           batch_size=_EVAL_BATCH_SIZE)
+
+  #mirrored_strategy = tf.distribute.MirroredStrategy()
+  # with mirrored_strategy.scope():
+  bert_layer = hub.KerasLayer(_BERT_LINK, trainable=False)
```

Comment:
I wonder what the best approach here is. Cuz like discussed, maybe it's best to freeze the weights for sometime then unfreeze them.
But if we go with this API, I'm not sure how to unfreeze certain layer in training.
And I kept it this way because training is a bit faster.